### ISSUE TITLE
vkQueuePresentKHR and semaphores

### ISSUE THREAD

**Issue by konstantinegorov** [KhronosGroup/Vulkan-Docs#1533]
> Queueing an image for presentation defines a set of queue operations, including waiting on the semaphores and submitting a presentation request to the presentation engine. However, the scope of this set of queue operations does not include the actual processing of the image by the presentation engine.

> if the presentation request is rejected by the presentation engine with an error VK_ERROR_OUT_OF_DATE_KHR, VK_ERROR_FULL_SCREEN_EXCLUSIVE_MODE_LOST_EXT, or VK_ERROR_SURFACE_LOST_KHR, the set of queue operations **are still considered to be enqueued and thus any semaphore wait operation specified in VkPresentInfoKHR will execute when the corresponding queue operation is complete.**

Is the following valid:

```
queue_submit(signal_semaphore: s)
queue_present(wait_semaphore: s) // VK_ERROR_OUT_OF_DATE_KHR
queue_submit(frame_fence)

wait_for_fence(frame_fence)
destroy_semaphore(s) // Is it safe? Does the frame_fence signal guarantee that the semaphore is no longer in use?
```

and:

```
queue_submit(signal_semaphore: s)
queue_present(wait_semaphore: s) // VK_ERROR_OUT_OF_DATE_KHR
queue_submit(signal_semaphore: s) // Is it safe?

```

**Comment by krOoze** [comment]
> `destroy_semaphore(s)`

Issue #152 (five years old now; time does fly...). @cubanismo ping here and the original issue.

>```
>queue_submit(signal_semaphore: s)
>queue_present(wait_semaphore: s) // VK_ERROR_OUT_OF_DATE_KHR
>queue_submit(signal_semaphore: s) // Is it safe?
>```

In the common practice this is avaided as probably invalid. Present (and sparse bind) is not part of queue submission order, so there's no dependency. They are considered somewhat as External Semaphores.

Annoyingly though you know when sparse bind finishes, but it is more tricky to infer for present. My assumption is that semaphore is ready for reuse if and only if the Presented image was seen again with Acquire. But that presents the problem that the semaphore is undestroyable in some cases (such as the `VK_ERROR_OUT_OF_DATE_KHR`); back to Issue #152.

Would be nice to finally have an official clarification.

**Comment by konstantinegorov** [comment]
Am I to understand that it's currently impossible to resize swapchain without a memory leak?

**Comment by stonesthrow** [comment]
https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/chap33.html#_wsi_swapchain - Search this for VK_ERROR_OUT_OF_DATE, there are several details. Basically destroy your out of date Swapchain, when the images are no longer being used for rendering, and create a new one that has the updated properties (size or whatever).

**Comment by TomOlson** [comment]
@konstantinegorov does @stonesthrow's suggestion address your issue? If so please close this, else let's continue the discussion.

**Comment by stonesthrow** [comment]
I will be following up with some word smithing to add clarity.

**Comment by stonesthrow** [comment]
> Am I to understand that it's currently impossible to recreate swapchain (due to VK_ERROR_OUT_OF_DATE) without a memory leak?

You are either destroying the swapchain, via vkDestroySwapchain, or using as "oldSwapchain" when creating a replacement vkCreateSwapchain. There can be only one swapchain associated with a surface at a time. And the images of the old swapchain must be returned to the Presentation Engine (QueuePresent). They will be OUT_OF_DATE. There is no leak because you are returning write ownership that you acquired with vkAcquireNextImage, and there is only one swapchain for a surface, so the old is destroyed. Once you get ERROR_OUT_OF_DATE, that swapchain is no longer valid for its Images to be presented, it will always return ERROR_OUT_OF_DATE. 

There are some clarifications in the works about the semaphore state when you get a failure like this.

**Comment by stonesthrow** [comment]
Can we close this issue?

**Comment by konstantinegorov** [comment]
Yes, It's duplicate of  #152

**Comment by krOoze** [comment]
@ShabbyX @konstantinegorov 152 was sorta closed, so this should probably too. I moved outstanding problems of that to #2007. Is there anything here to salvage?

**Comment by egorov-k** [comment]
There's another one question: what to do in the core version, without surface/swapchain maintenance extensions, but it was raised in the #2007, so this issue can be closed.

**Comment by oddhack** [comment]
This should be fixed in the 1.3.239 spec update.
